### PR TITLE
Improve WhiteHole documentation

### PR DESCRIPTION
**Pull request by MetaLang**
- Add Params section.
- Simplify the complex wording.
- Add LREF to the See_Also entries.
- Change example into a documented unittest.

### CODE REVIEW

File: std/typecons.d [hunk 1]
```diff
@@ -2638,33 +2638,36 @@ unittest
 
 /**
 $(D WhiteHole!Base) is a subclass of $(D Base) which automatically implements
-all abstract member functions as throw-always functions.  Each auto-implemented
-function fails with throwing an $(D Error) and does never return.  Useful for
-trapping use of not-yet-implemented functions.
+all abstract member functions as functions that always fail. These automatically
+implemented function simply throw an $(D Error) and never return. `Whitehole` is
```
Comment (Orvid) [inline std/typecons.d]: "implemented function" -> "implemented functions". Currently you're using both singular and plural in the same sentence.

File: std/typecons.d [hunk 2]
```diff
@@ -2638,33 +2638,36 @@ unittest
...
+{
+    import std.exception: assertThrown;
+
+    static class C
+    {
+        abstract void notYetImplemented();
+    }
+
+    auto c = new WhiteHole!C;
+    assertThrown!Throwable(c.notYetImplemented()); // throws an Error
```
Comment (Orvid) [inline std/typecons.d]: This should probably be asserting that `Error` is thrown, rather than `Throwable`.

File: std/typecons.d [hunk 3]
```diff
@@ -2638,33 +2638,36 @@ unittest
...
-}
-
-void main()
-{
-    auto c = new WhiteHole!C;
-    c.notYetImplemented(); // throws an Error
-}
---------------------
+Params:
+    Base = A non-final class from which `WhiteHole` can inherit.
```
Comment (Orvid) [inline std/typecons.d]: "can inherit" or "will inherit"?

File: std/typecons.d (hunk 1 above)
Comment (Orvid) [inline std/typecons.d]: Also, as we're trying to make it easier to understand, there's no reason to specify that these functions are automatically implemented, we've already done that in the previous sentence.

File: std/typecons.d (hunk 3 above)
Comment (MetaLang) [inline std/typecons.d]: "Can inherit" as in "has the ability to inherit from".

File: std/typecons.d (hunk 1 above)
Comment (MetaLang) [inline std/typecons.d]: Whoops, first one was a typo.

File: std/typecons.d (hunk 3 above)
Comment (Orvid) [inline std/typecons.d]: Then perhaps "A non-final class which `WhiteHole` is able to inherit from" would be better? With its current phrasing, it's possible to interpret it as "might inherit". (that's actually how I first interpreted it)

Comment (MetaLang) [inline std/typecons.d]: Somewhere, an English teacher just died ;-) I'll change the wording.

**Comment by MetaLang** [comment]
Fixed the nits, waiting for autotester.

**Comment by Orvid** [comment]
LGTM.

File: std/typecons.d [hunk 4]
```diff
@@ -2638,33 +2638,36 @@ unittest
...
+{
+    import std.exception: assertThrown;
+
+    static class C
+    {
+        abstract void notYetImplemented();
+    }
+
+    auto c = new WhiteHole!C;
+    assertThrown!Error(c.notYetImplemented()); // throws an Error
```
Comment (JakobOvrum) [inline std/typecons.d]: Not `NotImplementedError`? It is a documented exception.

Comment (MartinNowak) [inline std/typecons.d]: Can you please change that, otherwise good to go.

**Comment by MetaLang** [comment]
Nits fixed, ready to merge.

**Pull request merged**